### ISSUE TITLE
Generate PHP with namespace

### ISSUE THREAD

**Issue by tbillington** [google/protobuf#2741]
In the previous PHP grpc generator (dr slump) it was possible to [add a prefix to the namespace of the generated classes via an option in the proto files called `namespace`](
http://drslump.github.io/Protobuf-PHP/protobuf-php.5.html). Is this possible in the new official generator? I can't find docs on how to use it.

I use it to avoid naming conflicts with generated code and user facing code in a library.

If not, is there any way (outside of search replace on generated code) to do this with protoc?

Related: https://github.com/grpc/grpc/issues/9278

**Comment by TeBoring** [comment]
The recommended way is to use package in proto file.

**Comment by tbillington** [comment]
The issue is that there's not an easy way to submodule/subpackage or anything in php/composer land. I can't just put it in a different directory and get namespace separation with the rest of the library. I could search replace the root of the namespace on the generated code but that's quite ugly.

If I change the package in proto for php it will then break when it tries to communicate with other languages running the proto who have the original package.

Any ideas on how to do this 😛  ?

Basically, if my php library is called Foo, and the proto package starts with Foo, then the generated proto is basically surface/top level namespace in my library.

**Comment by TeBoring** [comment]
I think you can generate all proto files into a "generated" foler. In your composer file, you can add mapping from the root namespace to "generated":  https://getcomposer.org/doc/04-schema.md#psr-4
"":"generated"

**Comment by TeBoring** [comment]
Feel free to reopen the issue if it's not solved.

**Comment by mkaminski1988** [comment]
+1 Please add a prefix feature. We generate client code for java and go, which have options for each language's respective namespacing/packaging scheme. We're in a situation now where it's necessary to modify each protobuf's package value just to suit code generation for PHP.

**Comment by tbillington** [comment]
@mkaminski1988 seems it was implemented in grpc 1.6 https://github.com/grpc/grpc/releases/tag/v1.6.0
